### PR TITLE
fix(curriculum): updated tests to fix attribute sensitivity order

### LINKED ISSUE #52788: Step 31 of Build a Music Player is too sensitive to attribute order
### Describe the Issue

The [test code](https://github.com/freeCodeCamp/freeCodeCamp/blob/fb1f763357cc369d6ec517f74c098c08af5fc1ad/curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md?plain=1#L19) for step 31 of the "Build a Music Player" project requires a specific order to the `class` and `onclick` attributes. HTML (and JSX) do not require any particular ordering to attributes, so putting the `onclick` attribute before the `class` attribute should be acceptable.

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-basic-string-and-array-methods-by-building-a-music-player/step-31

### Your code

```jsx
      <button onclick="playSong(${song.id})" class="playlist-song-info">
          <span class="playlist-song-title">${song.title}</span>
          <span class="playlist-song-artist">${song.artist}</span>
          <span class="playlist-song-duration">${song.duration}</span>
      </button>
```

### Expected behavior

It should pass the tests.

### Screenshots

image

### System

- Device: ChromeBook
- OS: ChromeOS
- Browser: Chrome
- Version: 119.0.6045.212

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rakicodes**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #52788

### CODE REVIEW

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md [hunk 1]
```diff
@@ -13,10 +13,22 @@ Don't forget you need to interpolate with the dollar sign here.
 
 # --hints--
 
-You should add an `onclick` with the value `playSong(${song.id})` as the second attribute and value of the `button` element.
+You do not need to change the other parts of the code. All you need to do is add the `onclick` attribute to the button element.
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md:16]: I'm not crazy about this sentence. I think if we fix the tests so the order doesn't matter we can just simplify the original sentence.

```suggestion
You should add an `onclick` attribute with the value `playSong(${song.id})` to the `button` element.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md [hunk 2]
```diff
@@ -13,10 +13,22 @@ Don't forget you need to interpolate with the dollar sign here.
...
 ```js
-assert.match(code, /<button\s+class=('|")playlist-song-info\1\s*onclick=\1playSong\(\$\{song\.id\}\)\1>\s*<span\s+class=\1playlist-song-title\1>\$\{song\.title\}<\/span>\s*<span\s+class=\1playlist-song-artist\1>\$\{song\.artist\}<\/span>\s*<span\s+class=\1playlist-song-duration\1>\$\{song\.duration\}<\/span>\s*<\/button>/)
+assert.match(code, /<button\s.*>\s*<span\s+class=('|")playlist-song-title\1>\$\{song\.title\}<\/span>\s*<span\s+class=\1playlist-song-artist\1>\$\{song\.artist\}<\/span>\s*<span\s+class=\1playlist-song-duration\1>\$\{song\.duration\}<\/span>\s*<\/button>/)
+```
+
+You do not need to change the `class` attribute. All you need to do is add the `onclick` attribute to the button element.
+
+```js
+assert.match(code, /\s*class=('|")playlist-song-info\1\s*/)
+```
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md:26]: Not sure we need this.

**Review by lasjorg — CHANGES_REQUESTED** [review]

**Comment by lasjorg** [comment]
Do we even need regex here?

I think something like this should work the same in all browsers but I only tested it in Chrome and Firefox.

```
const button = document.querySelector('.playlist-song-info');
const handler = button.onclick.toString().replace(/\s+/g, '');
assert.equal(handler, 'functiononclick(event){playSong(0)}')
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md [hunk 3]
```diff
@@ -13,10 +13,22 @@ Don't forget you need to interpolate with the dollar sign here.
 
 # --hints--
 
-You should add an `onclick` with the value `playSong(${song.id})` as the second attribute and value of the `button` element.
+You do not need to change the other parts of the code. All you need to do is add the `onclick` attribute to the button element.
 
 ```js
-assert.match(code, /<button\s+class=('|")playlist-song-info\1\s*onclick=\1playSong\(\$\{song\.id\}\)\1>\s*<span\s+class=\1playlist-song-title\1>\$\{song\.title\}<\/span>\s*<span\s+class=\1playlist-song-artist\1>\$\{song\.artist\}<\/span>\s*<span\s+class=\1playlist-song-duration\1>\$\{song\.duration\}<\/span>\s*<\/button>/)
+assert.match(code, /<button\s.*>\s*<span\s+class=('|")playlist-song-title\1>\$\{song\.title\}<\/span>\s*<span\s+class=\1playlist-song-artist\1>\$\{song\.artist\}<\/span>\s*<span\s+class=\1playlist-song-duration\1>\$\{song\.duration\}<\/span>\s*<\/button>/)
```
Comment (ShaunSHamilton) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md:19]: @jdwilkin4 Are we not able to test the document object? E.g. `assert.includes(document.querySelector("button").classList, "playlist-song-title")`.

Comment (jdwilkin4) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md:19]: > Are we not able to test the document object?

I wasn't aware there was an issue with that. That should work

**Review by jdwilkin4 — COMMENTED** [review]
Hey @rakicodes ! 

Are you still interested in working on this PR?
It looks like there are a couple of things to resolve and then this PR should be ready for another review and approval.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md (hunk 1 above)
Comment (lasjorg) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md:16]: I think all we care about here is that the handler was added correctly. Then we can test that the button element still has the class on it in the next test.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md (hunk 3 above)
Comment (lasjorg) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-string-and-array-methods-by-building-a-music-player/65672136535209761a5cf02b.md:19]: Testing that everything remains the same seems out of scope for this step. If we had to do that every time our tests would become bloated.

The test message also becomes unclear, "something that shouldn't change, changed" isn't a very helpful message. The regex is just too large for a single assert message and it won't provide useful feedback.

I think all we do here is check that the class remains on the element after the handler is added.

**Comment by jdwilkin4** [comment]
HI @rakicodes ! 

Another PR was just approved and merged in which resolves the issue.

As a result, we are going to close this PR 👍

**Pull request closed without merging**